pkg/volume/csi/csi_attacher.go

```diff
@@ -76,52 +77,53 @@ func (c *csiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string
 	node := string(nodeName)
 	attachID := getAttachmentName(pvSrc.VolumeHandle, pvSrc.Driver, node)
 
-	var vaSrc storage.VolumeAttachmentSource
-	if spec.InlineVolumeSpecForCSIMigration {
-		// inline PV scenario - use PV spec to populate VA source.
-		// The volume spec will be populated by CSI translation API
-		// for inline volumes. This allows fields required by the CSI
-		// attacher such as AccessMode and MountOptions (in addition to
-		// fields in the CSI persistent volume source) to be populated
-		// as part of CSI translation for inline volumes.
-		vaSrc = storage.VolumeAttachmentSource{
-			InlineVolumeSpec: &spec.PersistentVolume.Spec,
-		}
-	} else {
-		// regular PV scenario - use PV name to populate VA source
-		pvName := spec.PersistentVolume.GetName()
-		vaSrc = storage.VolumeAttachmentSource{
-			PersistentVolumeName: &pvName,
-		}
+	attachment, err := c.plugin.volumeAttachmentLister.Get(attachID)
```

Comment:
I checked with someone on API machinery who suggested the staleness of data is usually on the order of hundreds of milliseconds, but in worst case scenarios it could be over a second.

In what scenario do you think this staleness could cause a problem?

Reply:
I do not know much about API machinery. Please ignore if you have already checked with an expert in this area. My scenario is:  Let’s say vol1 is attached to node1 and is then detached and is being attached now. Is it possible for VALister to return success because of stale cache?

Reply:
Yes it would be possible for the lister to return stale data which could cause it to think the VolumeAttachment already exists when it was just deleted. In this scenario the request would time out and then be retried next reconciler loop.

In the other race condition scenario where the VolumeAttachment doesn't show up in the lister (yet) but exists in the API, we ignore any "AlreadyExists" errors returned by the API and proceed with polling for the object until it's status shows a successful attachment.

Reply:
if you are concern the case that VA still exist, but volume is already detached, I think the following verifyVolumeAttachment will check VA deletionTimeStamp is already set, and will return with error